### PR TITLE
fix(archive): Only delete offloaded data we do not need. Fixes #2170 and #2156

### LINKED ISSUE #2170: Offload GC deletes offloaded data it should not
Checklist:

 * [ ] I've included the version.
 * [x] I've included reproduction steps.
 * [x] I've included the workflow YAML.
 * [x] I've included the logs.
 
**What happened**:

Offloaded nodes were deleted from the database, but the wf was still in Kubernetes.

```
time="2020-02-04T22:58:56Z" level=info msg="Deleting old UIDs that are not live" len_old_uids=2 len_wfs=0
time="2020-02-04T22:58:56Z" level=debug msg="Deleting offloaded nodes" uid=c9f481ed-2cd6-49aa-a539-54c382fc632f
```

**What you expected to happen**:

Not that.

**How to reproduce it (as minimally and precisely as possible)**:

Create a workflow that is offloaded and wait 5m.

### PR DESCRIPTION
**Pull request by alexec**
Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x ] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlike to be merged.
* [x] Optional. I've added My organization is added to the README.
* [x] I've signed the CLA and required builds are green. 

Fixes #2170 and #2156

### CODE REVIEW

File: workflow/controller/controller.go [hunk 1]
```diff
@@ -259,27 +259,31 @@ func (wfc *WorkflowController) periodicWorkflowGarbageCollector(stopCh <-chan st
...
 				}
-				if len(oldUIDs) == 0 {
-					log.Info("Zero old UIDs, nothing to do")
+				if len(oldRecords) == 0 {
+					log.Info("Zero old records, nothing to do")
 					return
 				}
-				list, err := util.NewWorkflowLister(wfc.wfInformer).List()
+				// we assume that this returns all records
+				list, err := wfc.wfclientset.ArgoprojV1alpha1().Workflows(wfc.GetManagedNamespace()).List(metav1.ListOptions{})
```
Comment (sarabala1979) [inline workflow/controller/controller.go]: Querying all available WFs every 5 mins will have a performance impact.

File: persist/sqldb/offload_node_status_repo.go [hunk 2]
```diff
@@ -166,32 +166,29 @@ func (wdc *nodeOffloadRepo) List(namespace string) (map[UUIDVersion]wfv1.Nodes,
 	return res, nil
 }
 
-func (wdc *nodeOffloadRepo) ListOldUIDs(namespace string) ([]string, error) {
+func (wdc *nodeOffloadRepo) ListOldOffloads(namespace string) ([]UUIDVersion, error) {
 	log.WithFields(log.Fields{"namespace": namespace}).Debug("Listing old offloaded nodes")
-	row, err := wdc.session.
-		Query("select distinct uid from "+wdc.tableName+" where clustername = ? and namespace = ? and updatedat < current_timestamp - interval '5' minute", wdc.clusterName, namespace)
+	var records []UUIDVersion
+	err := wdc.session.
```
Comment (alexec) [inline persist/sqldb/offload_node_status_repo.go]: we should have an index on (clustername,namespace,updatedat)

File: workflow/controller/controller.go [hunk 3]
```diff
@@ -259,27 +259,37 @@ func (wfc *WorkflowController) periodicWorkflowGarbageCollector(stopCh <-chan st
...
-				list, err := util.NewWorkflowLister(wfc.wfInformer).List()
-				if err != nil {
-					log.WithField("err", err).Error("Failed to list workflows")
-				}
-				wfs := make(map[types.UID]bool)
-				for _, wf := range list {
-					wfs[wf.UID] = true
+				// get every lives workflow (1000s) into a map
+				liveOffloadNodeStatusVersions := make(map[types.UID]string)
+				for cont := ""; ; {
```
Comment (sarabala1979) [inline workflow/controller/controller.go]: Still, I feel we are doing the costly operation here. list all workflows and iterate them. I am trying to see we can optimize list operations to get with UID. So We just need to make very few calls based on `ListOldOffloads`. I know this PR is a show stopper for v2.5 release. I would like @jessesuen to take a quick look and suggest a better way.

File: ui/src/app/workflows/components/workflow-details/workflow-details.tsx [hunk 4]
```diff
@@ -334,7 +334,8 @@ export class WorkflowDetails extends React.Component<RouteComponentProps<any>, {
     private async loadWorkflow(namespace: string, name: string) {
         try {
             this.ensureUnsubscribed();
-            this.changesSubscription = services.workflows.watch({name, namespace})
+            this.changesSubscription = services.workflows
```
Comment (alexec) [inline ui/src/app/workflows/components/workflow-details/workflow-details.tsx]: linting

File: workflow/controller/controller.go [hunk 5]
```diff
@@ -259,27 +259,31 @@ func (wfc *WorkflowController) periodicWorkflowGarbageCollector(stopCh <-chan st
...
-				oldUIDs, err := wfc.offloadNodeStatusRepo.ListOldUIDs(wfc.GetManagedNamespace())
+				oldRecords, err := wfc.offloadNodeStatusRepo.ListOldOffloads(wfc.GetManagedNamespace())
 				if err != nil {
-					log.WithField("err", err).Error("Failed to list offloaded nodes")
+					log.WithField("err", err).Error("Failed to list old offloaded nodes")
 				}
-				if len(oldUIDs) == 0 {
-					log.Info("Zero old UIDs, nothing to do")
+				if len(oldRecords) == 0 {
+					log.Info("Zero old records, nothing to do")
```
Comment (sarabala1979) [inline workflow/controller/controller.go]: minor,  Can you reverse the logging if there are records? 
```
if len(oldRecords) > 0 {
```
which will reduce the logging

**Pull request merged**